### PR TITLE
chore(eslint): upgrade to eslint@4

### LINKED ISSUE #870: ESlint v4
Hi! Today is the official release of eslint v4.0.0. `eslint --init` can break because `eslint-plugin-import` only support eslint v3.x.

```
? How would you like to configure ESLint? Use a popular style guide
? Which style guide do you want to follow? Standard
? What format do you want your config file to be in? YAML
Local ESLint installation not found.
Installing eslint, eslint-plugin-standard, eslint-plugin-promise, eslint-config-standard
...
├─┬ eslint@4.0.0
├── UNMET PEER DEPENDENCY eslint-plugin-import@>=2.2.0
```

### PR DESCRIPTION
**Pull request by mastilver**
close #870

### CODE REVIEW

**Comment by mastilver** [comment]
The failing tests are also failing on `master`, I will try to find what's going on

I would like to get some help to apply the same logic for appveyor as I did for travis! :)

**Comment by ShayDavidson** [comment]
Hey, any update on this? Thanks!

**Comment by Arcanemagus** [comment]
@mastilver For AppVeyor you will need to put each of the 6 combinations as individual groups in the environment matrix, so something like this:

```yml
environment:
  matrix:
    - nodejs_version: "7"
      ESLINT_VERSION: "4"

    - nodejs_version: "7"
      ESLINT_VERSION: "3"

    - nodejs_version: "7"
      ESLINT_VERSION: "2"

    - nodejs_version: "6"
      ESLINT_VERSION: "4"

    - nodejs_version: "6"
      ESLINT_VERSION: "3"

    - nodejs_version: "6"
      ESLINT_VERSION: "2"
```

Note also that it seems npm v3 is being forcibly installed for some reason, possibly an artifact from when Node.js v0.10 was still supported. As this is a downgrade for both of those versions of Node.js this should probably be removed unless it is still testing something relevant.

**Comment by ljharb** [comment]
npm 3 is absolutely still supported, as is npm 2, because that's what node 4 and 6 ship with.

That said, we shouldn't be forcibly installing an older version of npm.

**Comment by mastilver** [comment]
@evilebottnawi Yeah, I know... :)

I need to allocate more time for it, I will continue tonight or Sunday
In the meantime if someone want to take over feel free or ping me and I'll add you to the fork

File: appveyor.yml [hunk 1]
```diff
@@ -1,8 +1,23 @@
 # Test against this version of Node.js
 environment:
   matrix:
-  - nodejs_version: "7"
-  - nodejs_version: "6"
+    - nodejs_version: "8"
+      ESLINT_VERSION: "4"
```
Comment (ljharb) [inline appveyor.yml]: looks like the `_` might be messing this one up

Comment (mastilver) [inline appveyor.yml]: I think the issue is that I'm still using the linux syntax... `npm install eslint@$ESLINT_VERSION` 🤦‍♂️

**Comment by mastilver** [comment]
Any windows / appveyor expert here?

I tried without success:
- `npm install eslint@$(eslint_version)`
- `npm install eslint@${eslint_version}`
- `npm install eslint@${env:eslint_version}`
- `npm install eslint@$env:eslint_version`

**Comment by corbinu** [comment]
@mastilver I believe it is `%eslint_version%`

**Comment by lukeapage** [comment]
Seems its now failing with 

```
npm run coveralls
> eslint-plugin-import@2.4.0 coveralls C:\projects\eslint-plugin-import
> nyc report --reporter lcovonly && cat ./coverage/lcov.info | coveralls
[error] "2017-06-19T10:40:53.216Z"  'error from getOptions'
C:\projects\eslint-plugin-import\node_modules\coveralls\bin\coveralls.js:18
        throw err;
        ^
Error: Command failed: git cat-file -p a35349b81b811d23df7f3387a664d06937f743da
fatal: Not a valid object name a35349b81b811d23df7f3387a664d06937f743da
    at ChildProcess.exithandler (child_process.js:252:12)
    at emitTwo (events.js:125:13)
```

as explained here:

https://github.com/nickmerwin/node-coveralls/issues/80#issuecomment-147487373 

Not sure why
```
git clone -q https://github.com/benmosher/eslint-plugin-import.git C:\projects\eslint-plugin-import
git fetch -q origin +refs/pull/871/merge:
git checkout -qf FETCH_HEAD
```
would fail to clone everything needed though

File: tests/src/rules/named.js [hunk 2]
```diff
@@ -135,7 +135,6 @@ ruleTester.run('named', rule, {
 
     test({
       code: 'import { a } from "./re-export-names"',
-      args: [2, 'es6-only'],
```
Comment (ljharb) [inline tests/src/rules/named.js]: why need this line be removed?

File: tests/src/rules/default.js [hunk 3]
```diff
@@ -58,8 +58,7 @@ ruleTester.run('default', rule, {
 
     // #94: redux export of execution result,
     test({ code: 'import connectedApp from "./redux"' }),
-    test({ code: 'import App from "./jsx/App"'
-         , ecmaFeatures: { jsx: true, modules: true } }),
```
Comment (ljharb) [inline tests/src/rules/default.js]: these should be passed inside `parserOptions`, instead of removed.

File: src/rules/no-commonjs.js [hunk 4]
```diff
@@ -41,7 +41,7 @@ module.exports = {
 
       },
       'CallExpression': function (call) {
-        if (context.getScope().type !== 'module') return
+        if (context.getScope().type !== 'module' && context.getScope().type !== 'global') return
```
Comment (ljharb) [inline src/rules/no-commonjs.js]: why is this change needed?

File: src/rules/no-amd.js [hunk 5]
```diff
@@ -17,7 +17,7 @@ module.exports = {
         return {
 
             'CallExpression': function (node) {
-          if (context.getScope().type !== 'module') return
+          if (context.getScope().type !== 'module' && context.getScope().type !== 'global') return
```
Comment (ljharb) [inline src/rules/no-amd.js]: why is this change needed?

**Comment by mastilver** [comment]
@ljharb Thank you for your review, you were right about `global` not being needed, it was just an issue with `test()` in `tests/src/utils.js`

**Comment by benmosher** [comment]
@ljharb your review concerns have been totally addressed, yes?

**Comment by benmosher** [comment]
btw @mastilver this build matrix of eslint + node versions is really great! a major upgrade!

**Comment by benmosher** [comment]
looks like the last failure is at https://github.com/benmosher/eslint-plugin-import/blob/master/tests/src/rules/named.js#L138

`args` => `options`, I think? reference: http://eslint.org/docs/developer-guide/working-with-rules

**Comment by benmosher** [comment]
also re: 

> Note also that it seems npm v3 is being forcibly installed for some reason, possibly an artifact from when Node.js v0.10 was still supported. As this is a downgrade for both of those versions of Node.js this should probably be removed unless it is still testing something relevant.

you are correct, it wouldn't install for testing correctly without npm v3. I don't think this is necessary anymore, but needs to be confirmed.

also, apologies: I merged an older PR updating CI to test on node 4/6/8. this may need to be updated as well.

FWIW I think it'd be okay to test fewer combinations on Windows but I'll defer to @ljharb's judgment on that. Appveyor is slow and unreliable so it might just add noise.

**Comment by mastilver** [comment]
> you are correct, it wouldn't install for testing correctly without npm v3. I don't think this is necessary anymore, but needs to be confirmed.

I'm only upgrading to npm@3 on node@4

> FWIW I think it'd be okay to test fewer combinations on Windows but I'll defer to @ljharb's judgment on that. Appveyor is slow and unreliable so it might just add noise.

Yeah, let me know what should I do, maybe remove eslint matrix for appveyor and just keep the node matrix?

**Comment by benmosher** [comment]
@mastilver
> remove eslint matrix for appveyor and just keep the node matrix
sounds good to me 👍🏻

**Comment by Arcanemagus** [comment]
Looks like the failing Travis CI build is on a test that normally takes a while to run, as that specific job was the longest running is that an intermittent failure on the slower OS X hosts?

File: tests/src/rules/named.js (hunk 2 above)
Comment (benmosher) [inline tests/src/rules/named.js]: oh actually, I think this line is dead code and can be removed. `es6-only` was a v1 option that was removed ages ago. 

it's fine if it's just `options` now though, can clean up later

Comment (benmosher) [inline tests/src/rules/named.js]: note how it's only used here: https://github.com/benmosher/eslint-plugin-import/search?utf8=✓&q=es6-only&type=

**Comment by mastilver** [comment]
@benmosher I think it would be handy to enable `Auto cancel pull request builds` on travis

**Comment by benmosher** [comment]
Ah, neat. I didn't even know that was an option, thanks! And great work! 🎉🙏🏻🎉

**Comment by lukeapage** [comment]
@benmosher please can we have a new release to unblock using eslint 4?

**Comment by ljharb** [comment]
@lukeapage https://github.com/benmosher/eslint-plugin-import/issues/870#issuecomment-310547554 please be patient. Using eslint 4 isn't urgent :-)

**Comment by lukeapage** [comment]
Sorry should have checked the issue, I thought it had just been forgotten.

**Comment by benmosher** [comment]
published 👍🏻
